### PR TITLE
Tsc 373 fix download of datapack error

### PR DESCRIPTION
**Pull request by JacquiFreeElectron**
fix the problem of uuid being undefined when requesting a download. Also, no file will be downloaded to user's computer if not logged in

https://github.com/earthhistoryviz/tsconline/assets/104879321/e0491008-a2ac-4725-a496-f149a83bff8f

### CODE REVIEW

File: app/src/state/actions/general-actions.ts [hunk 1]
```diff
@@ -733,7 +734,12 @@ export const requestDownload = action(async (filename: string, needEncryption: b
         );
         break;
       }
+      case 401: {
+        displayServerError(response, ErrorCodes.NOT_LOGGED_IN, ErrorMessages[ErrorCodes.NOT_LOGGED_IN]);
+        break;
+      }
```
Comment (paologum) [inline app/src/state/actions/general-actions.ts:740]: for less dupe code, and also for a default value to be in place, you should have the cases change the error code, and at the end of the `if`, `displayServerError` and set a default value initially, that way unknown status codes will still print a value

File: app/src/settings_tabs/Datapack.tsx [hunk 2]
```diff
@@ -44,17 +44,19 @@ export const Datapacks = observer(function Datapacks() {
...
+        await new Promise((resolve, reject) => {
+          reader.onloadend = resolve;
+          reader.onerror = reject;
+        });
+        return reader.result;
+      } catch (error) {
+        actions.pushError(ErrorCodes.INVALID_PATH);
+        return "";
+      }
     }
```
Comment (paologum) [inline app/src/settings_tabs/Datapack.tsx:60]: if `getFileURL` returns null or empty, `handleDownload` creates an `aTag` that never gets removed, so you should either always remove it or only create it if `fileURL` exists

**Review by paologum — CHANGES_REQUESTED** [review]

File: app/src/settings_tabs/Datapack.tsx (hunk 2 above)
Comment (JacquiFreeElectron) [reply app/src/settings_tabs/Datapack.tsx:60]: I think `handleDownload `only create `aTag` when `fileURL` exists bc I checked `if (fileURL && typeof fileURL === "string") `

File: app/src/state/actions/general-actions.ts [hunk 3]
```diff
@@ -709,31 +709,31 @@ export const requestDownload = action(async (filename: string, needEncryption: b
...
-          ErrorCodes.INCORRECT_ENCRYPTION_HEADER,
-          ErrorMessages[ErrorCodes.INCORRECT_ENCRYPTION_HEADER]
-        );
+      case 401: {
+        errorCode = ErrorCodes.NOT_LOGGED_IN;
+        errorMessage = ErrorMessages[ErrorCodes.NOT_LOGGED_IN];
         break;
       }
     }
+    displayServerError(response, errorCode, errorMessage);
```
Comment (paologum) [inline app/src/state/actions/general-actions.ts:735]: errormessage is just `ErrorMessages[errorCode]` so you don't need to set it each time, you can just set `errorCode` then abstract it here

**Review by paologum — APPROVED** [review]

**Pull request merged**